### ISSUE TITLE
Optimise PyTorch Performance

### ISSUE THREAD

**Issue by danielkelshaw** [MagriLab/PISR#10]
There are a number of things we can to improve the performance of training, for instance:

- [x] Asynchronous data-loading / pinned memory
- [x] Non-blocking transfer to GPU
- [x] Set gradients to None rather than zero.
- [x] Disable gradient accumulation for validation.
- [x] Use `torch.backends.cudnn.benchmark = True` to optimise CNN performance.
- [x] Shuffle data to avoid 'bad batches' and allow better generalisation.
- [x] Increase the batch-size as much as possible.

**Note:** if we use `torch.backends.cudnn.benchmark = True` then we should drop the last batch in the dataloader as this will often be a different size to the rest of the batches - this will incur additional computational cost as benchmarks are run.
